Objects/floatobject.c

```diff
@@ -2002,29 +1998,30 @@ _PyFloat_InitTypes(PyInterpreterState *interp)
 }
 
 void
-_PyFloat_ClearFreeList(PyInterpreterState *interp)
+_PyFloat_ClearFreeList(_PyFreeListState *freelist_state, int is_finalization)
 {
-#if PyFloat_MAXFREELIST > 0
-    struct _Py_float_state *state = &interp->float_state;
+#ifdef WITH_FREELISTS
+    struct _Py_float_state *state = &freelist_state->float_state;
     PyFloatObject *f = state->free_list;
     while (f != NULL) {
         PyFloatObject *next = (PyFloatObject*) Py_TYPE(f);
         PyObject_Free(f);
         f = next;
     }
     state->free_list = NULL;
-    state->numfree = 0;
+    if (is_finalization) {
```

Comment:
Previously, we only set `num_free` to `-1` during finalization in debug builds. With this change, we set it in non-debug builds too. I think that's fine (and simpler), but we should better handle the `-1` case in `PyFloat_ExactDealloc` in case assertions are not enabled. 

I'd suggest adding changing the condition in `PyFloat_ExactDealloc` to something like `if (state->num_free >= PyFloat_MAXFREELIST || state->num_free < 0)` so that `num_free=-1` disables the free-list.